providers/flagsmith/src/main/java/dev.openfeature.contrib.providers.flagsmith/FlagsmithProvider.java

```diff
@@ -0,0 +1,232 @@
+     */
+    private <T> ProviderEvaluation<T> buildEvaluation(
+        T flagValue, ErrorCode errorCode, Reason reason, String variationType) {
+        ProviderEvaluation.ProviderEvaluationBuilder providerEvaluationBuilder =
+            ProviderEvaluation.<T>builder()
+                              .value(flagValue);
+
+        if (errorCode != null) {
+            providerEvaluationBuilder.errorCode(errorCode);
+        }
+        if (reason != null) {
+            providerEvaluationBuilder.reason(reason.name());
+        }
+        if (variationType != null) {
+            providerEvaluationBuilder.variant(variationType);
+        }
+
+        return providerEvaluationBuilder.build();
+    }
+
+    /**
+     * The method convertValue is converting the object return by the Flagsmith client.
+     *
+     * @param value        the value we have received
+     * @param expectedType the type we expect for this value
+     * @param <T>          the type we want to convert to
+     * @return A converted object
+     */
+    private <T> T convertValue(Object value, Class<?> expectedType) {
+        boolean isPrimitive = expectedType == Boolean.class
+            || expectedType == String.class
+            || expectedType == Integer.class
+            || expectedType == Double.class;
+
+        if (isPrimitive) {
+            return (T) value;
+        }
+        return (T) objectToValue(value);
+    }
+
+    /**
+     * The method objectToValue is wrapping an object into a Value.
+     *
+     * @param object the object you want to wrap
+     * @return the wrapped object
+     */
+    private Value objectToValue(Object object) {
+        if (object instanceof Value) {
+            return (Value) object;
+        } else if (object == null) {
+            return null;
+        } else if (object instanceof String) {
+            return new Value((String) object);
+        } else if (object instanceof Boolean) {
+            return new Value((Boolean) object);
+        } else if (object instanceof Integer) {
+            return new Value((Integer) object);
+        } else if (object instanceof Double) {
+            return new Value((Double) object);
+        } else if (object instanceof Structure) {
+            return new Value((Structure) object);
+        } else if (object instanceof List) {
+            // need to translate each elem in list to a value
+            return new Value(((List<Object>) object).stream()
+                                                    .map(this::objectToValue)
+                                                    .collect(Collectors.toList()));
+        } else if (object instanceof Instant) {
+            return new Value((Instant) object);
+        } else if (object instanceof Map) {
+            return new Value(mapToStructure((Map<String, Object>) object));
+        } else {
+            throw new TypeMismatchError("Flag value " + object + " had unexpected type "
```

Comment:
seems like we're throwing this in the method above as well, do we need it in both places?

Reply:
This error is thrown when the object type is unknown wheras the other is thrown when the type does not match the one expected.